### PR TITLE
Iden3 integration

### PR DESCRIPTION
**Pull request by KyrylR**

### CODE REVIEW

File: contracts/iden3/State.sol
Comment (Arvolear) [inline contracts/iden3/State.sol:1]: Is this used?

File: contracts/Voting.sol [hunk 1]
```diff
@@ -1,99 +1,213 @@
...
+        votingInfo.values.commitmentStartTime = votingParams_.commitmentStart;
+        votingInfo.values.votingStartTime =
+            votingParams_.commitmentStart +
+            votingParams_.commitmentPeriod;
+        votingInfo.values.votingEndTime =
+            votingInfo.values.votingStartTime +
+            votingParams_.votingPeriod;
+
+        for (uint256 i = 0; i < votingParams_.candidates.length; i++) {
+            candidates[votingParams_.candidates[i]] = true;
```
Comment (Arvolear) [inline contracts/Voting.sol:86]: How many candidates are planned?

File: contracts/Voting.sol [hunk 2]
```diff
@@ -1,99 +1,213 @@
...
         rootsHistory[getRoot()] = true;
 
-        emit UserRegistered(commitment_, block.number);
+        emit UserRegistered(msg.sender, proveIdentityParams_, registerProofParams_, block.number);
     }
 
+    /**
+     * @inheritdoc IVoting
+     */
     function vote(
```
Comment (Arvolear) [inline contracts/Voting.sol:133]: There are no commitment checks?

File: contracts/VotingFactory.sol [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+        _register(votingType_, msg.sender, voting_);
+
+        emit VotingCreated(votingType_, msg.sender, voting_);
+    }
+
+    /**
+     * @inheritdoc IVotingFactory
+     */
+    function createVoting(
```
Comment (Arvolear) [inline contracts/VotingFactory.sol]: Let's use different (not overloaded) naming

File: contracts/Voting.sol (hunk 2 above)
Comment (Arvolear) [inline contracts/Voting.sol:136]: What is the relationship between commitment and nullifier? I see no circom circuits here

File: test/Voting.test.ts [hunk 4]
```diff
@@ -102,121 +218,124 @@ describe("Voting", () => {
...
+      ).to.be.revertedWith("Voting: Invalid vote proof");
     });
-  });
 
-  describe("#addRoot", () => {
-    it("should record a root in history only by the owner", async () => {
-      await expect(voting.connect(FIRST).addRoot(ethers.ZeroHash))
-        .to.be.revertedWithCustomError(voting, "OwnableUnauthorizedAccount")
-        .withArgs(FIRST.address);
+    it("should revert if vote with incorrect nullifier hash", async () => {
```
Comment (Arvolear) [inline test/Voting.test.ts]: Misleading test name

File: contracts/iden3/State.sol
Comment (KyrylR) [reply contracts/iden3/State.sol:1]: Yep, in the deployment scripts

File: contracts/Voting.sol (hunk 2 above)
Comment (KyrylR) [reply contracts/Voting.sol:136]: Same as always: 
H(Secret | Nullifier)  = Commitment 
H(Nullifier) = Nullifier Hash

Repository with relevant circuits: https://github.com/rarimo/passport-zk-circuits

File: contracts/iden3/validators/QueryValidator.sol [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    function _getInputValidationParameters(
+        uint256[] memory inputs_
+    ) internal pure virtual returns (ValidationParams memory);
+
+    function _checkGistRoot(uint256 gistRoot_) internal view {
+        ILightweightState.GistRootData memory rootData_ = lightweightState.geGISTRootData(
+            gistRoot_
+        );
+
+        require(
```
Comment (Arvolear) [inline contracts/iden3/validators/QueryValidator.sol:96]: Can this be frontrun?

File: contracts/iden3/LightweightState.sol [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        GistRootData calldata gistData_,
+        bytes calldata proof_
+    ) external override {
+        _checkMerkleSignature(_getSignHash(gistData_, newIdentitiesStatesRoot_), proof_);
+
+        require(
+            !isIdentitiesStatesRootExists(newIdentitiesStatesRoot_),
+            "LightweightState: Identities states root already exists"
+        );
+        require(
```
Comment (Arvolear) [inline contracts/iden3/LightweightState.sol:76]: This may revert on identical gist roots

File: contracts/iden3/verifiers/BaseVerifier.sol [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+    function _checkAllowedIssuer(string memory queryId_, uint256 issuerId_) internal view virtual {
+        require(
+            isAllowedIssuer(zkpQueriesStorage.getStoredSchema(queryId_), issuerId_),
+            "BaseVerifier: Issuer is not on the list of allowed issuers."
+        );
+    }
+
+    function _checkChallenge(uint256 challenge_) internal view virtual {
+        require(
```
Comment (Arvolear) [inline contracts/iden3/verifiers/BaseVerifier.sol]: msg.sender is the voting contract here

File: contracts/VotingRegistry.sol [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+import {Paginator} from "@solarity/solidity-lib/libs/arrays/Paginator.sol";
+
+import {IVotingRegistry} from "./interfaces/IVotingRegistry.sol";
+
+/**
+ * @title VotingRegistry contract
+ */
+contract VotingRegistry is IVotingRegistry, Initializable, OwnableUpgradeable {
```
Comment (Arvolear) [inline contracts/VotingRegistry.sol]: Other contracts are UUPS. For consistency reasons I would suggest rewriting those to be UUPS as well

File: package.json [hunk 9]
```diff
@@ -21,17 +21,21 @@
     "lint-sol-fix": "prettier --write \"contracts/**/*.sol\""
   },
   "dependencies": {
-    "@openzeppelin/contracts": "5.0.1",
-    "@openzeppelin/contracts-upgradeable": "5.0.1",
-    "@solarity/solidity-lib": "2.6.15"
+    "@iden3/contracts": "1.0.0",
+    "@openzeppelin/contracts": "4.9.5",
+    "@openzeppelin/contracts-upgradeable": "4.9.5",
+    "@rarimo/evm-bridge-contracts": "1.0.0",
```
Comment (Arvolear) [inline package.json]: Update this to the latest version please

File: contracts/Voting.sol (hunk 2 above)
Comment (Arvolear) [inline contracts/Voting.sol:136]: Is it ok if dude revokes his claim but still votes?

File: contracts/iden3/LightweightState.sol [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        );
+        bytes32 computedRoot_ = MerkleProof.processProofCalldata(
+            statesMerkleData_.merkleProof,
+            merkleLeaf_
+        );
+
+        return (isIdentitiesStatesRootExists(computedRoot_), computedRoot_);
+    }
+
+    function _authorizeUpgrade(
```
Comment (Arvolear) [inline contracts/iden3/LightweightState.sol:149]: ?

File: contracts/Voting.sol [hunk 11]
```diff
@@ -1,99 +1,213 @@
...
+        } else {
+            registerVerifier.proveRegistration(proveIdentityParams_, registerProofInfo_);
+        }
+
         _add(commitment_);
         commitments[commitment_] = true;
         rootsHistory[getRoot()] = true;
 
-        emit UserRegistered(commitment_, block.number);
+        emit UserRegistered(msg.sender, proveIdentityParams_, registerProofParams_, block.number);
```
Comment (Arvolear) [inline contracts/Voting.sol]: block.number?

File: contracts/Voting.sol (hunk 2 above)
Comment (KyrylR) [reply contracts/Voting.sol:136]: Yep

File: contracts/Voting.sol (hunk 1 above)
Comment (KyrylR) [reply contracts/Voting.sol:86]: Added an upper limit of 100 candidates

File: contracts/VotingFactory.sol (hunk 3 above)
Comment (KyrylR) [reply contracts/VotingFactory.sol]: Done, changed names to avoid collisions

File: test/Voting.test.ts (hunk 4 above)
Comment (KyrylR) [reply test/Voting.test.ts]: Changed the test name

File: contracts/iden3/validators/QueryValidator.sol (hunk 5 above)
Comment (KyrylR) [reply contracts/iden3/validators/QueryValidator.sol:96]: Nope, as it works for previous roots too (it does not rely on the current root only)

File: contracts/iden3/LightweightState.sol (hunk 6 above)
Comment (KyrylR) [reply contracts/iden3/LightweightState.sol:76]: The method which calls the `signedTransitState` method does validation

File: contracts/iden3/verifiers/BaseVerifier.sol (hunk 7 above)
Comment (KyrylR) [reply contracts/iden3/verifiers/BaseVerifier.sol]: Deleted the _checkChallenge check in favor of the onlyVoting modifier

File: contracts/VotingRegistry.sol (hunk 8 above)
Comment (KyrylR) [reply contracts/VotingRegistry.sol]: Done

File: package.json (hunk 9 above)
Comment (KyrylR) [reply package.json]: Done

File: contracts/iden3/LightweightState.sol (hunk 10 above)
Comment (KyrylR) [reply contracts/iden3/LightweightState.sol:149]: It is intended logic

File: contracts/Voting.sol (hunk 11 above)
Comment (KyrylR) [reply contracts/Voting.sol]: Deleted

File: test/helpers/iden3/utils.ts [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+  // proof of non-inclusion (NodeAux does not exist)
+  return {
+    key: 0n,
+    value: 0n,
+    noAux: 1n,
+  };
+}
+
+// // PoseidonHashValue returns the solidity and circom implementation of poseidon hash
+// func PoseidonHashValue(values []*big.Int) (*big.Int, error) {
```
Comment (Arvolear) [inline test/helpers/iden3/utils.ts]: Redundant code

**Review by Arvolear — APPROVED** [review]
LGTM!

**Pull request merged**